test/SILOptimizer/noimplicitcopy_consuming_parameters.swift

```diff
@@ -0,0 +1,993 @@
+}
+
+func testLoadableConsumingNonEscapingClosure(_ x: consuming NonTrivialStruct) {
+    func useNonEscaping(_ f: () -> ()) {}
+    useNonEscaping {
+        _ = x
+    }
+}
+
+func testLoadableConsumingNonEscapingClosure2(_ x: consuming NonTrivialStruct) {
+    // expected-error @-1 {{'x' used after consume}}
+    let _ = x // expected-note {{consumed here}}
+    func useNonEscaping(_ f: () -> ()) {}
+    useNonEscaping { // expected-note {{used here}}
+        _ = x
+    }
+}
+
+func testLoadableConsumingConsumeOperator(_ x: consuming NonTrivialStruct) {
+    _ = consume x
+}
+
+func testLoadableConsumingConsumeOperator1(_ x: consuming NonTrivialStruct) {
+    // expected-error @-1 {{'x' used after consume}}
+    _ = consume x // expected-note {{consumed here}}
+    borrowValDefault(x) // expected-note {{used here}}
+}
+
+func testLoadableConsumingConsumeOperator2(_ x: consuming NonTrivialStruct) {
+    x = consume x
+}
+
+func testLoadableConsumingConsumeOperator3(_ x: consuming NonTrivialStruct) {
+    _ = consume x
+    let y = NonTrivialStruct()
+    x = consume y
+}
+
+func testLoadableConsumingEnum(_ x: consuming LoadableEnum) {
+    switch x {
+    case let .x(y):
+        _ = y
+        break
+    case .y:
+        break
+    }
+}
+
+func testLoadableConsumingEnum2(_ x: consuming LoadableEnum) {
+    // expected-error @-1 {{'x' consumed more than once}}
+    switch x { // expected-note {{consumed here}}
+    case let .x(y):
+        _ = consume x // expected-note {{consumed again here}}
+        _ = y
+        break
+    case .y:
+        break
+    }
+}
+
+func testLoadableConsumingCopyOperator(_ x: consuming NonTrivialStruct) {
+    _ = copy x
+}
+
+//////////////////////////////////////////
+// MARK: Trivial Struct Consuming Tests //
+//////////////////////////////////////////
+
+func testTrivialConsumingConsume(_ x: consuming TrivialStruct) {
+    // expected-error @-1 {{'x' used after consume}}
+    let _ = x // expected-note {{consumed here}}
+    let _ = x.k // expected-note {{used here}}
+}
+
+func testTrivialConsumingConsume2(_ x: consuming TrivialStruct) {
+    // expected-error @-1 {{'x' consumed more than once}}
+    var y = x // expected-note {{consumed here}}
+    y = x // expected-note {{consumed again here}}
+    _ = y
+}
+
+func testTrivialConsumingConsume3(_ x: consuming TrivialStruct) {
+    let y = x
+    _ = y
+}
+
+func testTrivialConsumingUse(_ x: consuming TrivialStruct) {
+    _ = x
+}
+
+func testTrivialConsumingUseAndConsume(_ x: consuming TrivialStruct) {
+    borrowValBorrowing(x)
+    let _ = x
+}
+
+func testTrivialConsumingConsumeAndUse(_ x: consuming TrivialStruct) {
+    // expected-error @-1 {{'x' used after consume}}
+    let _ = x // expected-note {{consumed here}}
+    borrowValBorrowing(x) // expected-note {{used here}}
+}
+
+func testTrivialConsumingConsumeAndUseReinit(_ x: consuming TrivialStruct) {
+    let _ = x
+    x = TrivialStruct()
+    borrowValBorrowing(x)
+}
```

Comment:
something I haven't seen yet is that we're still preventing implicit copies on the parameter binding even after you've reassigned it, whether it's a `consuming` or `borrowing` parameter. Also makes me wonder what happens when you do `x = x`... does it require you to write `x = copy x` ?

Also, a test that entures we still enforce no-implicit-copy after an explicit copy of the param.